File: csharp/src/code-smells/couplers/FeatureEnvy.cs
namespace RefactoringAvanzado.CodeSmells.Couplers;

public class Customer
{
    public string Name { get; }
    public string Street { get; }
    public string City { get; }
    public string Zip { get; }

    public Customer(string name, string street, string city, string zip)
    {
        Name = name;
        Street = street;
        City = city;
        Zip = zip;
    }
}

public class ShippingCalculator
{
    public int Cost(Customer customer)
    {
        var baseAmount = customer.Zip.StartsWith("9") ? 10 : 20;
        var distant = customer.City.Length > 6 ? 5 : 0;
        return baseAmount + distant;
    }
}

public static class FeatureEnvyDemo
{
    public static int DemoFeatureEnvy(Customer c)
    {
        return new ShippingCalculator().Cost(c);
    }
}

File: typescript/src/code-smells/couplers/feature-envy.ts
export class Customer {
  constructor(
    public name: string,
    public street: string,
    public city: string,
    public zip: string,
  ) {
  }
}

export class ShippingCalculator {
  cost(customer: Customer): number {
    const base = customer.zip.startsWith('9') ? 10 : 20
    const distant = customer.city.length > 6 ? 5 : 0
    return base + distant
  }
}

export function demoFeatureEnvy(c: Customer): number {
  return new ShippingCalculator().cost(c)
}
